frame/examples/default-config/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+		/// reliant on `frame_system::Config` or the overarching runtime in any way.
+		type WithDefaultValue: Get<u32>;
+
+		/// Same as [`Config::WithDefaultValue`], but we don't intend to define a default for this
+		/// in our tests below.
+		type OverwrittenDefaultValue: Get<u32>;
+
+		/// An input parameter that relies on `<Self as frame_system::Config>::AccountId`. As of
+		/// now, such types cannot have defaults and need to be annotated as such, iff
+		/// `#[pallet::config(with_default)]` is enabled:
+		#[pallet::no_default]
+		type CannotHaveDefault: Get<Self::AccountId>;
+
+		/// Something that is a normal type, with default.
+		type WithDefaultType;
+
+		/// Same as [`Config::WithDefaultType`], but we don't intend to define a default for this
+		/// in our tests below.
+		type OverwrittenDefaultType;
+	}
+
+	/// Container for different types that implement [`DefaultConfig`]` of this pallet.
+	pub mod config_preludes {
+		// This will help use not need to disambiguate anything when using `derive_impl`.
+		use super::*;
+
+		/// A type providing default configurations for this pallet in testing environment.
+		pub struct TestDefaultConfig;
+		#[frame_support::register_default_impl(TestDefaultConfig)]
+		impl DefaultConfig for TestDefaultConfig {
+			type WithDefaultValue = frame_support::traits::ConstU32<42>;
+			type OverwrittenDefaultValue = frame_support::traits::ConstU32<42>;
+
+			type WithDefaultType = u32;
+			type OverwrittenDefaultType = u32;
+		}
+
+		/// A type providing default configurations for this pallet in a parachain environment.
+		pub struct ParachainDefaultConfig;
+		#[frame_support::register_default_impl(ParachainDefaultConfig)]
+		impl DefaultConfig for ParachainDefaultConfig {
+			type WithDefaultValue = frame_support::traits::ConstU32<66>;
+			type OverwrittenDefaultValue = frame_support::traits::ConstU32<66>;
+			type WithDefaultType = u32;
+			type OverwrittenDefaultType = u32;
+		}
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	pub enum Event<T: Config> {}
+}
+
+#[cfg(any(test, doc))]
+pub mod tests {
+	use super::*;
+
+	use frame_support::macro_magic::use_attr;
+	// Because `derive_impl` is a [macro_magic](https://crates.io/crates/macro_magic) attribute
+	// macro, [`#[use_attr]`](`frame_support::macro_magic::use_attr`) must be attached to any use
+	// statement that brings it into scope.
+	#[use_attr]
+	use frame_support::derive_impl;
+
+	use super::pallet as pallet_default_config_example;
+
+	type UncheckedExtrinsic = frame_system::mocking::MockUncheckedExtrinsic<Test>;
+	type Block = frame_system::mocking::MockBlock<Test>;
+
+	frame_support::construct_runtime!(
+		pub enum Test where
+			Block = Block,
+			NodeBlock = Block,
+			UncheckedExtrinsic = UncheckedExtrinsic,
+		{
+			System: frame_system,
+			DefaultPallet: pallet_default_config_example,
+		}
+	);
+
+	#[derive_impl(frame_system::config_preludes::TestDefaultConfig as frame_system::DefaultConfig)]
```

Comment:
This `as DefaultConfig` syntax still confuses me, in the sense that I have no idea why it's necessary and what it does. It would just seem to me that most of time, you can just omit it.

Reply:
You can omit it in certain situations indeed. It is documented here https://github.com/paritytech/substrate/pull/13454/files#diff-55fa2bbcb98405e03a4311fca1d14b211001133cffa61fb0d4837be69c9e7bebR804